Janitorial: clean up deprecated code

jeherve opened the pull request:
#### Changes proposed in this Pull Request:

* Following #16433, let's remove all code that was deprecated more than 6 months ago.

I left out everything in the compat package.

#### Jetpack product discussion

* N/A

#### Does this pull request change what data or activity we track or use?

* No

#### Testing instructions:

* Do tests pass? 
* There is a lot of code that was touched, so I'd recommend looking at different Jetpack features, the connection process, as well as the matching WordPress.com diff and whether those functions / methods are still used somewhere on WordPress.com.

kraftbj reviewed (dismissed):
I'd like to see about running this realize against the Analyzer to see what might break. This would be a good thing to check it with. :)

jeherve commented:
> I'd like to see about running this realize against the Analyzer to see what might break

Yes, that'd be nice, but last time I tried I ran out of memory and could not get it to work. @zinigor Do you still have your setup to analyze against all plugins in the directory? Do you think you could run it with this PR?

zinigor commented:
Sure thing, I'll run it today and report my findings here.

jeherve commented:
Thank you!

zinigor commented:
I ran the analyzer, here's the results that it found in the updated WordPress plugins repo:

```
# Akismet
method_missing,class.akismet-admin.php,1130,"Class static method \Jetpack::load_xml_rpc_client() is missing","\Jetpack::load_xml_rpc_client()"

# Certly Guard
method_missing,class.certly-admin.php,948,"Class static method \Jetpack::load_xml_rpc_client() is missing","\Jetpack::load_xml_rpc_client()"

# Jetpack Subscription Form
method_missing,jetpack-subscription.php,253,"Class static method \Jetpack::load_xml_rpc_client() is missing","\Jetpack::load_xml_rpc_client()"

# Reflect
function_missing,reflect.php,93,"Function migrate() is missing",migrate()
function_missing,reflect.php,142,"Function migrate() is missing",migrate()

# SlimJetpack
method_missing,class.jetpack-cli.php,93,"Class static method \Jetpack::unlink_user($user_id=\null) is missing","\Jetpack::unlink_user($user_id=\null)"
method_missing,class.jetpack-client.php,87,"Class static method \Jetpack::fix_url_for_bad_hosts($url) is missing","\Jetpack::fix_url_for_bad_hosts($url)"
method_missing,modules/videopress/videopress.php,154,"Class static method \Jetpack::load_xml_rpc_client() is missing","\Jetpack::load_xml_rpc_client()"
method_missing,class.jetpack.php,1431,"Class static method \Jetpack::clean_nonces($all=\false) is missing","\Jetpack::clean_nonces($all=\false)"
method_missing,class.jetpack.php,1433,"Class static method \Jetpack::load_xml_rpc_client() is missing","\Jetpack::load_xml_rpc_client()"
method_missing,class.jetpack.php,1469,"Class static method \Jetpack::load_xml_rpc_client() is missing","\Jetpack::load_xml_rpc_client()"
method_missing,class.jetpack.php,1554,"Class static method \Jetpack::fix_url_for_bad_hosts($url) is missing","\Jetpack::fix_url_for_bad_hosts($url)"
method_missing,class.jetpack.php,1554,"Class static method \Jetpack::api_url($relative_url) is missing","\Jetpack::api_url($relative_url)"
method_missing,class.jetpack.php,2754,"Class static method \Jetpack::api_url($relative_url) is missing","\Jetpack::api_url($relative_url)"
method_missing,class.jetpack.php,3115,"Class static method \Jetpack::load_xml_rpc_client() is missing","\Jetpack::load_xml_rpc_client()"
method_missing,class.jetpack.php,3135,"Class static method \Jetpack::load_xml_rpc_client() is missing","\Jetpack::load_xml_rpc_client()"
method_missing,class.jetpack.php,3302,"Class static method \Jetpack::fix_url_for_bad_hosts($url) is missing","\Jetpack::fix_url_for_bad_hosts($url)"
method_missing,class.jetpack.php,3302,"Class static method \Jetpack::api_url($relative_url) is missing","\Jetpack::api_url($relative_url)"
method_missing,class.jetpack.php,3594,"Class static method \Jetpack::load_xml_rpc_client() is missing","\Jetpack::load_xml_rpc_client()"
method_missing,class.jetpack.php,3634,"Class static method \Jetpack::load_xml_rpc_client() is missing","\Jetpack::load_xml_rpc_client()"
method_missing,class.jetpack.php,3881,"Class static method \Jetpack::load_xml_rpc_client() is missing","\Jetpack::load_xml_rpc_client()"
method_missing,class.jetpack-ixr-client.php,15,"Class static method \Jetpack::xmlrpc_api_url() is missing","\Jetpack::xmlrpc_api_url()"
method_missing,class.jetpack-client-server.php,79,"Class static method \Jetpack::update_user_token($user_id,$token,$is_master_user) is missing","\Jetpack::update_user_token($user_id,$token,$is_master_user)"
method_missing,class.jetpack-client-server.php,170,"Class static method \Jetpack::fix_url_for_bad_hosts($url) is missing","\Jetpack::fix_url_for_bad_hosts($url)"
method_missing,class.jetpack-client-server.php,170,"Class static method \Jetpack::api_url($relative_url) is missing","\Jetpack::api_url($relative_url)"
method_missing,class.jetpack-debugger.php,88,"Class static method \Jetpack::fix_url_for_bad_hosts($url) is missing","\Jetpack::fix_url_for_bad_hosts($url)"

# VaultPress
method_missing,vaultpress.php,2999,"Class static method \Jetpack::load_xml_rpc_client() is missing","\Jetpack::load_xml_rpc_client()"
method_missing,vaultpress.php,3018,"Class static method \Jetpack::load_xml_rpc_client() is missing","\Jetpack::load_xml_rpc_client()"

$ Widget for EventBrite API
function_missing,frontend/class-frontend.php,65,"Function jetpack_photon_parse_url($url,$component=-1) is missing","jetpack_photon_parse_url($url,$component=-1)"
```

The `migrate` thing is weird though, I don't know where it came from. To sanity check these results here's what the difference to the latest Jetpack version:

```
method_missing,modules/comments.php,27,"Class static method \Jetpack::dns_prefetch($new_urls=\null) is missing","\Jetpack::dns_prefetch($new_urls=\null)"
method_missing,modules/likes.php,16,"Class static method \Jetpack::dns_prefetch($new_urls=\null) is missing","\Jetpack::dns_prefetch($new_urls=\null)"
method_missing,modules/masterbar/masterbar.php,116,"Class static method \Jetpack::dns_prefetch($new_urls=\null) is missing","\Jetpack::dns_prefetch($new_urls=\null)"
method_missing,modules/photon-cdn.php,17,"Class static method \Jetpack::dns_prefetch($new_urls=\null) is missing","\Jetpack::dns_prefetch($new_urls=\null)"
method_missing,modules/comment-likes.php,16,"Class static method \Jetpack::dns_prefetch($new_urls=\null) is missing","\Jetpack::dns_prefetch($new_urls=\null)"
method_missing,modules/shortcodes/videopress.php,13,"Class static method \Jetpack::dns_prefetch($new_urls=\null) is missing","\Jetpack::dns_prefetch($new_urls=\null)"
method_missing,modules/photon.php,15,"Class static method \Jetpack::dns_prefetch($new_urls=\null) is missing","\Jetpack::dns_prefetch($new_urls=\null)"
method_missing,class.jetpack-gutenberg.php,76,"Class static method \Jetpack_Gutenberg::register_plugin($slug) is missing","\Jetpack_Gutenberg::register_plugin($slug)"
method_missing,class.jetpack-gutenberg.php,94,"Class static method \Jetpack_Gutenberg::set_extension_unavailability_reason($slug,$reason) is missing","\Jetpack_Gutenberg::set_extension_unavailability_reason($slug,$reason)"
```

jeherve commented:
Putting this on hold since we'll need to make changes to Akismet and VaultPress first.

brbrr commented:
Akismet does check Jetpack version: `if ( defined( 'JETPACK__VERSION' ) && version_compare( JETPACK__VERSION, '7.7', '<' )  ) {`

https://plugins.trac.wordpress.org/browser/akismet/trunk/class.akismet-admin.php#L1129

Vaultpress also checks for class availability:

```

// For version of Jetpack prior to 7.7.
if ( ! class_exists( 'Jetpack_IXR_Client' ) ) {
Jetpack::load_xml_rpc_client();
}

```
https://plugins.trac.wordpress.org/browser/vaultpress/trunk/vaultpress.php#L2998

So I'd say we safe merging this.

@jeherve Do you think we should remove these conditionals any way?

zinigor commented:
There's a problem with the VaultPress check. It only checks for the class existence, not for the method. The class still does exist, and it exists in newer Jetpack versions too. I haven't checked, but I'm pretty sure this flow will cause a fatal error with this PR.

brbrr commented:
It does make sense. It seems that we would need to update the conditional to check for jetpack version similar to how akismet is doing it now. And possibly remove it completely after most vaultpress users do update to latest Jetpack version

brbrr commented:
I made a PR to fix the conditional in Vaultpress: https://github.com/Automattic/vaultpress/pull/100

jeherve commented:
@zinigor Do you think you could run this against the Analyzer for the whole plugin repo, once again?

Thank you!

cc @sergeymitr since we talked about this during our team chat :)

gibrown commented on modules/search/class.jetpack-search.php:
```diff
@@ -1440,21 +1440,6 @@ public function set_filters( array $aggregations ) {
-	 *
-	 * @param array $facets Array of facets to apply to the search.
-	 */
-	public function set_facets( array $facets ) {
```
@nickdaugherty do you think this is safe for all VIPs to have removed? My guess is yes. I think it is a holdover from the old WPCOM_elasticsearch code, but maybe worth searching through all VIP repos. I don't see anything on wpcom.

nickdaugherty replied:
With GH's search I was able to find 1 instance of a VIP site using this directly, we're doing a more comprehensive check now.

nickdaugherty replied:
Looks like this is indeed the only instance, we're working with the customer to get it switched.

kraftbj replied:
Thanks @nickdaughtery. In terms of timeline, the second Tuesday of January is when this version would go stable (presuming it lands this month). We're pushing a week later due to the holiday season.

nickdaugherty replied:
Thanks @kraftbj - we'll be all set on our end by that date, so 👍 from me on this change.

kraftbj commented:
This fell off my radar but let's get it in. A little more time has passed but I'm fine keeping it as is (e.g. not adding a couple more months of cleanup).

jeherve committed:
Janitorial: clean up deprecated code

Following #16433, let's remove all code that was deprecated more than 6 months ago.
This commit also standardizes some of the version numbers so they can start getting the proper notices, and we can then remove the deprecations in the future.

Note that "6 months" here refers to 6 months as of the last time this PR was updated, in December 2020.

jeherve committed:
Add changelogs

jeherve committed:
Remove more code that was deprecated more than 6 months ago

jeherve committed:
Remove endpoints that are not used anymore

jeherve committed:
Remove deprecated filter

jeherve committed:
Update test comments to refer to updated function name

jeherve committed:
Stop requiring deleted file

Also add blank file back to avoid issues with sites with strong caching mechanisms.

jeherve committed:
Remove tests for removed methods

jeherve committed:
Remove deprecated test

jeherve committed:
Remove unused methods

bluefuton commented:
@gibrown does it feel safe to remove these old methods from the Search class? They were deprecated in 5.0:

https://github.com/Automattic/jetpack/pull/16434/files#diff-6f42dd4b44978d4b744cad6222469cec24cf3f6dad9b50e9b7848201e94216fb

jeherve committed:
Merge remote-tracking branch 'origin/master' into rm/deprecated-code

kraftbj commented on projects/plugins/jetpack/class.jetpack.php:
```diff
@@ -5573,178 +5164,6 @@ public static function maybe_set_version_option() {
-	 *
-	 * @deprecated since 7.7.0
-	 */
-	public static function load_xml_rpc_client() {
```
This function is called by the following plugins in the w.org repo:

- [ ] `akismet`: class.akismet-admin.php,1136
- [ ] `certly-guard`: class.certly-admin.php,948
- [ ] `jetpack-subscription-form`: jetpack-subscription.php,253
- [ ] `vaultpress`: vaultpress.php,3004 and vaultpress.php,3023

kraftbj replied:
All good:
- `akismet`: Wrapped in `if ( defined( 'JETPACK__VERSION' ) && version_compare( JETPACK__VERSION, '7.7', '<' )  ) {` ([#](https://plugins.trac.wordpress.org/browser/akismet/tags/4.1.10/class.akismet-admin.php#L1136))
- `vaultpress`: Both instances wrapped in `if ( defined( 'JETPACK__VERSION' ) && version_compare( JETPACK__VERSION, '7.7', '<' ) && ! class_exists( 'Jetpack_IXR_Client' ) ) {` ([#](https://plugins.trac.wordpress.org/browser/vaultpress/trunk/vaultpress.php#L3003))

Would break things:

- `jetpack-subscription-form` This plugin "adds customizable UI" to our subscription form. Last updated 5 years ago, 500+ active installs. It calls this function without any type of check (would fatal if active with Jetpack deactivated). https://wordpress.org/plugins/jetpack-subscription-form/

Would break, but don't care :)
- `certly-guard` -- I didn't check this one. It has under 10 active installs and requires an API Key from a defunct site..

kraftbj commented on projects/plugins/jetpack/class.jetpack.php:
```diff
@@ -6200,19 +5568,6 @@ public static function check_identity_crisis() {
-	 * @since  3.8.0
-	 * @return bool True = already allowed False = not on the allowed list.
-	 */
-	public static function is_staging_site() {
```
This function is called by the following plugins in the w.org repo:

- [ ] `connect-for-woocommerce`: classes/class-wc-connect-api-client.php,409
- [ ] `connect-for-woocommerce`: classes/class-wc-connect-help-view.php,129
- [ ] `easyreservations`: includes/class-er-tracker.php,105
- [ ] `geodirectory`: includes/admin/class-geodir-admin-tracker.php,114
- [ ] `woocommerce`: includes/class-wc-tracker.php,108
- [ ] `woocommerce-services`: classes/class-wc-connect-jetpack.php,72

kraftbj replied:
- `connect-for-woocommerce` -- Old a8c plugin that is closed. only 1 outdated JP install with it. Going to mark this one fine.
- `easyreservations` -- Same as Woo below. New way, then deprecated with proper gating.
- `geodirectory` -- [Uses the old way, but has a proper gate](https://plugins.trac.wordpress.org/browser/geodirectory/trunk/includes/admin/class-geodir-admin-tracker.php#L113). **We should reach out to encourage them to update. Under active development.**
- `woocommerce` -- Will return the new way first and does verify that `is_staging_site` is a callable before using it worst case.
- `woocommerce-services` -- Same as above. New way first, the deprecated as a backup for old versions with a proper check.

kraftbj replied:
`geodirectory` -- Left message https://wordpress.org/support/topic/update-jetpack-function-2/

kraftbj commented on projects/plugins/jetpack/modules/minileven.php:
```diff
@@ -1,54 +0,0 @@
-/**
- * Check if we are on mobile.
- */
-function jetpack_check_mobile() {
```
This function is used by the following w.org plugins:

- [ ] `gator-cache`: gator-cache.php,676

kraftbj replied:
`gator-cache` last updated 3 years ago, 600+ installs. 

`jetpack_check_mobile` [is gated so it will only work](https://plugins.trac.wordpress.org/browser/gator-cache/trunk/gator-cache.php#L671) if `minileven` is in the active module array. I don't immediately recall if we removed that from the active module array when we deprecated minileven. If we did, we can clear this one out.

kraftbj replied:
Confirmed that [we will deactivate deprecated modules](https://github.com/Automattic/jetpack/blob/5605eb656d65149dbeebec31077bc13c8ea1d1a2/projects/plugins/jetpack/class.jetpack.php#L2579), so no modern sites should have `minileven` as an active module.

kraftbj commented on projects/plugins/jetpack/class.jetpack-user-agent.php:
```diff
@@ -5,660 +5,7 @@
-	 *
-	 * @deprecated 8.7.0 Use Automattic\Jetpack\Device_Detection\User_Agent_Info
-	 */
-	public static function is_ipad( $type = 'ipad-any' ) {
```
This function is used by the following w.org plugins:

- [ ] `infinite-transporter`: themes/twentyfourteen.php,30
- [ ] `jp-custom-css`: custom-css/custom-css.php,31
- [ ] `themedy-toolbox`: custom-css/custom-css/custom-css.php,26
- [ ] `wpvkp-ultimate-shortcodes`: custom-css/custom-css/custom-css.php,26

I have a feeling a couple of these are because places have incorporated our code into theirs, but should check it out.

kraftbj replied:
- `infinite-transporter` Gated, but in a way that doesn't help (`jetpack_is_mobile` is the gate). Last update 7 yrs ago, 90+ installs which seem to include ~20 Jetpack sites.
- `jp-custom-css` No longer maintained. Like the above, gated by `jetpack_is_mobile`. No longer available (closed plugin). Installed on 25 Jetpack sites.
- `themedy-toolbox`: Last update 3 years ago, 1000+ installs. Seems to be the same code as `jp-custom-css` with a `jetpack_is_mobile` gate. 140 Jetpack installs.
- `wpvkp-ultimate-shortcodes`: Closed plugin. 3 Jetpack installs. Same code as above, `jetpack_is_mobile` gate.

kraftbj commented on projects/plugins/jetpack/class.jetpack-user-agent.php:
```diff
@@ -5,660 +5,7 @@
- *
- * @deprecated 8.7.0 Use Automattic\Jetpack\Device_Detection\User_Agent_Info
  */
-class Jetpack_User_Agent_Info {
```
This class is used in the following plugins:

- [ ] `jetpack-omnisearch`: omnisearch/class.jetpack-user-agent.php,8
- [ ] `jp-custom-css`: custom-css/class.jetpack-user-agent.php,8
- [ ] `themedy-toolbox`: custom-css/custom-css/class.jetpack-user-agent.php,8
- [ ] `wpvkp-ultimate-shortcodes`: custom-css/custom-css/class.jetpack-user-agent.php,8

kraftbj replied:
`jetpack-omnisearch` is closed, 2 JP installs. This usage is fine (it brings in the user agent the same way as the ones above, with a `jetpack_is_mobile` check... yay for code sharing). Our tool didn't find that class actually being used though. Given the small impact, going to check this box off.

The other ones were mentioned above, they included the class so we can check it off for this instance.

kraftbj commented on projects/plugins/jetpack/class.jetpack-user-agent.php:
```diff
@@ -5,660 +5,7 @@
-	 *
-	 * @return boolean
-	 */
-	public static function is_bot_user_agent( $ua = null ) {
```
This function is used in the following w.org plugins:

- [ ] `jetpack-omnisearch`: omnisearch/class.jetpack-user-agent.php,1376
- [ ] `jp-custom-css`: custom-css/class.jetpack-user-agent.php,1376
- [ ] `themedy-toolbox`: custom-css/custom-css/class.jetpack-user-agent.php,1367
- [ ] `wpvkp-ultimate-shortcodes`: custom-css/custom-css/class.jetpack-user-agent.php,1367

kraftbj replied:
All of these instances are fine. They are bringing in the class (though with issues as noted above).

kraftbj commented on projects/plugins/jetpack/functions.photon.php:
```diff
@@ -305,24 +305,6 @@ function jetpack_photon_url_scheme( $url, $scheme ) {
- * @param integer $component Retrieve specific URL component.
- * @return mixed Result of parse_url
- */
-function jetpack_photon_parse_url( $url, $component = -1 ) {
```
This function is used in the following w.org plugin:

- [ ] `widget-for-eventbrite-api`: frontend/class-frontend.php,65

kraftbj replied:
`widget-for-eventbrite-api` is under active development (13hr ago last update) and 2k installs. Function is not gated ([#](https://plugins.trac.wordpress.org/browser/widget-for-eventbrite-api/trunk/frontend/class-frontend.php#L62)).

The function can be wholesale replaced with `wp_parse_url` and their min WP version works. Opened a support ticket: https://wordpress.org/support/topic/update-jetpack-function/#new-topic-0

Suggest we leave this in for now with adding an internal note to remove it a fair amount of time after they've updated their plugin.

kraftbj commented on projects/plugins/jetpack/modules/custom-css/custom-css.php:
```diff
@@ -1646,208 +1646,6 @@ static function filter_attr( $css, $element = 'div' ) {
 	}
 }
 
-function migrate() {
```
This function is called in the following plugin:

- [ ] `reflect` reflect.php,93 and 142

My guess is this is a badly named function on our part and they're declaring their own version.

kraftbj replied:
[This plugin defined a function of the same name](https://plugins.trac.wordpress.org/browser/reflect/trunk/reflect.php#L113). False positive.

kraftbj reviewed (changes requested):
A few things for us to check. The analyzer is not smart enough to tell if the call to these functions are within some kind of check that would make it okay (e.g. if JP version is under 7.7, run jetpack_load_xml...), so just need to manually check them.

If any are legit, we should reach out to those plugin authors if possible and, for now, keep the functions around.

jeherve committed:
Merge remote-tracking branch 'origin/master' into rm/deprecated-code

jeherve committed:
Add method back as it is used in a third-party plugin

see https://github.com/Automattic/jetpack/pull/16434/files#r667210696

jeherve committed:
Add Jetpack_User_Agent_Info back, with a note

Since multiple plugins are using it, let's not get rid of it just yet.

See https://github.com/Automattic/jetpack/pull/16434/files#r667190852

jeherve committed:
Still require file

Since we are not getting rid of the class just yet, we should still require it.

jeherve committed:
Bring back tests since we are keeping the class a bit longer.

jeherve committed:
Merge remote-tracking branch 'origin/master' into rm/deprecated-code

jeherve committed:
Add test provider back to avoid test failures

We'll just leave the test in, since it could be useful in the future for other file deprecations.

jeherve committed:
Merge remote-tracking branch 'origin/master' into rm/deprecated-code

jeherve commented:
r233300-wpcom

jeherve committed:
Update version number for connection package

jeherve commented:
@kraftbj I had to update a package's version number. Could I ask for one last review? :)

Thank you!
